src/providers/proxy/proxy_hosts.c

```diff
@@ -21,14 +21,308 @@
 */
 
 #include "providers/proxy/proxy.h"
+#include "db/sysdb_iphosts.h"
+#include <arpa/inet.h>
+
+static errno_t
+nss_status_to_errno(enum nss_status status)
+{
+    switch (status) {
+    case NSS_STATUS_SUCCESS:
+        return EOK;
+    case NSS_STATUS_TRYAGAIN:
+        return EAGAIN;
+    case NSS_STATUS_NOTFOUND:
+        return ENOENT;
+    case NSS_STATUS_UNAVAIL:
+    default:
+        break;
+    }
+
+    return EIO;
+}
+
+static errno_t
+parse_hostent(TALLOC_CTX *mem_ctx,
+              struct hostent *result,
+              bool case_sensitive,
+              char **out_name,
+              char ***out_aliases,
+              char ***out_addresses)
+{
+    char **addresses = *out_addresses;
+    char **aliases = *out_aliases;
+    int i;
+    errno_t ret;
+
+    /* Parse addresses */
+    for (i = 0; result->h_addr_list[i] != NULL; i++) {
+        size_t len = talloc_array_length(addresses);
+        char buf[INET6_ADDRSTRLEN];
+        const char *addr = NULL;
+        bool found = false;
+        int j;
+
+        if (result->h_length == INADDRSZ) {
+            addr = inet_ntop(AF_INET, result->h_addr_list[i],
+                             buf, INET6_ADDRSTRLEN);
+        } else if (result->h_length == IN6ADDRSZ) {
+            addr = inet_ntop(AF_INET6, result->h_addr_list[i],
+                             buf, INET6_ADDRSTRLEN);
+        }
+
+        if (addr == NULL) {
+            DEBUG(SSSDBG_MINOR_FAILURE,
+                    "Failed to convert host network address of host "
+                    "'%s' to a character string: %s\n", result->h_name,
+                    strerror(errno));
+            continue;
+        }
+
+        /* Skip duplicates */
+        for (j = 0; j < len && addresses[j] != NULL; j++) {
```

Comment:
There is a clang warning that `Array access (from variable 'addresses') results in a null pointer dereference` (same for `aliases` below as well). I think `j < len` should prevent this but maybe you can add a NULL check to make clang happy?